### DISCUSSION TITLE
Is there a way to switch databases?

### DISCUSSION THREAD

**Discussion by cemremengu** [go-jet/jet#187] [Q&A]
I want my application to support both sqlite and postgres. I am happy to include generated code for both databases in the final executable. 

However, is there a way to switch between implementations during the startup without rebuilding the binary? Any experience with this?

**Comment by go-jet** [comment]
Hi @cemremengu. 
I'm not sure what you mean. If the generated code for both databases is included in the build, then code will be include in the binary.

**Reply by mesonyktikon** [reply]
@go-jet I think I have the same question.

To use this code
```
stmt := SELECT(Users.Email).FROM(Users).WHERE(Users.Email.EQ(String(email)))
```

I include this import
```
. "github.com/go-jet/jet/v2/sqlite"
```

But let's say that I want to use sqlite just for my local environment and postgres for the remote environment. Doesn't this imply importing `github.com/go-jet/jet/v2/postgres` instead?

Both these packages provide `SELECT` and friends. It seems like a build time change to the imports is needed when deciding which db backend to use. OP seems to be asking is there's a runtime equivalent of making this change.

**Reply by go-jet** [reply]
Statements are independent of database connection/transaction. `Query` and `Exec` methods accept `Queryable` and `Executable` interfaces, meaning you can pass the DB (or Tx) object of any SQL database(Postgres, MySQL, SQLite, Oracle, ....).
So there is nothing stopping you from writing just statements for SQLite and then switching dynamically db connection on startup to sqlite or postgres database(or the other way around, writing just Postgres statements).   
There are slight differences between different database dialects and not every statement will run on other databases. In your case, `postgres` and `sqlite` dialect are quite similar, so most queries should work.
One important difference between `postgres` and `sqlite` generated sql is in schema name. `sqlite` does not support different schema, so they don't appear in generated sql, unless sqlite database is attached. You can overcome this issue by removing the schema name from all statements with [table.UsingSchema("")](https://github.com/go-jet/jet/wiki/FAQ#how-to-use-jet-in-multi-tenant-environment) global function.
So it is possible, although I don't understand why you would do it. Why not just run Postgres on your local environment as well?
